server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -512,9 +512,9 @@ public void updateShardState(final ShardRouting newRouting,
                                 engine.rollTranslogGeneration();
                                 engine.fillSeqNoGaps(newPrimaryTerm);
                                 if (getMaxSeqNoOfUpdatesOrDeletes() == SequenceNumbers.UNASSIGNED_SEQ_NO) {
-                                    // TODO: Enable this assertion after we replicate max_seq_no_updates during replication
-                                    // assert indexSettings.getIndexVersionCreated().before(Version.V_7_0_0_alpha1) :
-                                    //    indexSettings.getIndexVersionCreated();
+                                    // If the old primary was on an old version, this promoting primary (was replica before)
+                                    // does not have max_seq_no_of_updates. We need to bootstrap it manually from its local history.
+                                    assert indexSettings.getIndexVersionCreated().before(Version.V_7_0_0_alpha1);
                                     engine.initializeMaxSeqNoOfUpdatesOrDeletes();
```

Comment:
I think we need maxSeq no here? maybe the translog wasn't fsynced? (if people disable the per request fsync)

Reply:
Yes, initializeMaxSeqNoOfUpdatesOrDeletes uses the max_seq_no from the local tracker and translog.